smdebug/core/error_handler.py

```diff
@@ -0,0 +1,136 @@
+            the rest of training. A default return value is returned when an error is caught or a wrapped function is
+            called when smdebug is already disabled.
+
+            The default return value of the wraooed function (in the event of an error) must be specified in
+            `default_return_value` if it isn't valid for the wrapped function to return `None`.
+
+            If the default return value can only be determined at runtime (i.e. layer or tape callback), a function
+            can be provided in `default_return_value` and the default return value will be determined dynamically by
+            calling that function with the inputs provided to the wrapped function.
+
+            Currently, the error handler will only catch errors if the default smdebug configuration is being used.
+            Otherwise, the error will be raised normally. When an error is caught, the stack trace of the error will still
+            be logged for tracking purposes.
+
+            Examples:
+
+            ```
+            error_handler = SMDebugErrorHandler.get_error_handler()
+            ...
+            @error_handler.catch_smdebug_errors()
+            def foo(*args, **kwargs):
+                ...
+                return
+
+            @error_handler.catch_smdebug_errors(default_return_val=False)
+            def bar(*args, **kwargs):
+                ...
+                return True
+
+            def foobar(*args, **kwargs):
+                default_func = lambda *args, **kwargs: {"args": args, "kwargs": kwargs}
+
+                @error_handler.catch_smdebug_errors(default_return_val=default_func)
+                def baz()
+                    ...
+                return baz
+            ```
+            """
+
+            def wrapper(func):
+                @functools.wraps(func)
+                def error_handler(*args, **kwargs):
+                    # Return immediately if smdebug is disabled.
+                    if self.disable_smdebug:
+                        if callable(default_return_val):
+                            return default_return_val(*args, **kwargs)
+                        return default_return_val
+
+                    try:
+                        # Attempt calling the smdebug function and returning the output
+                        return func(*args, **kwargs)
+                    except Exception as e:
+                        # If an smdebug error occurred with the default configuration or it occurred before the
+                        # configuration can even be determined (i.e. the constructor), catch the error and log it.
+
+                        if self.hook is None or self.hook.has_default_configuration():
+                            self.logger.error(BASE_ERROR_MESSAGE)
+                            self.logger.exception(e)  # Log stack trace.
+                            self.disable_smdebug = True  # Disable smdebug
+
+                            if callable(default_return_val):
```

Comment:
Nice! I learned something new. Can you explain what happens here as a comment?

Reply:
Explained it in the docstring for the class, but will add a comment here too.